File: run-migration.js
/**
 * Database Migration Script
 * Runs the SQL migration files in the backend/db/migrations directory
 */

const fs = require('fs');
const path = require('path');
const { Pool } = require('pg');
const dotenv = require('dotenv');

// Load environment variables
dotenv.config();

// Database configuration
const dbConfig = {
  user: process.env.DB_USER || 'postgres',
  password: process.env.DB_PASSWORD || 'vishwak',
  host: process.env.DB_HOST || 'localhost',
  port: process.env.DB_PORT || 5432,
  database: process.env.DB_NAME || 'property_mapping',
  // Add connection timeout
  connectionTimeoutMillis: 5000
};

console.log('Running database migrations...');
console.log('Database config:', {
  user: dbConfig.user,
  host: dbConfig.host,
  database: dbConfig.database,
  port: dbConfig.port
});

// Create a connection pool
const pool = new Pool(dbConfig);

// Path to migration files
const migrationsPath = path.join(__dirname, 'backend', 'db', 'migrations');

// Function to run a migration file
async function runMigration(filePath) {
  const sql = fs.readFileSync(filePath, 'utf8');
  const client = await pool.connect();
  
  try {
    console.log(`Running migration: ${path.basename(filePath)}`);
    await client.query('BEGIN');
    await client.query(sql);
    await client.query('COMMIT');
    console.log(`✅ Migration ${path.basename(filePath)} completed successfully`);
    return true;
  } catch (error) {
    await client.query('ROLLBACK');
    console.error(`❌ Migration ${path.basename(filePath)} failed:`, error.message);
    return false;
  } finally {
    client.release();
  }
}

// Function to run all migrations
async function runMigrations() {
  try {
    // Create migrations table if it doesn't exist
    const client = await pool.connect();
    try {
      await client.query(`
        CREATE TABLE IF NOT EXISTS migrations (
          id SERIAL PRIMARY KEY,
          name VARCHAR(255) NOT NULL,
          applied_at TIMESTAMP DEFAULT NOW()
        )
      `);
    } finally {
      client.release();
    }
    
    // Get list of migration files
    const files = fs.readdirSync(migrationsPath)
      .filter(file => file.endsWith('.sql'))
      .sort(); // Sort to ensure migrations run in order
    
    // Get already applied migrations
    const result = await pool.query('SELECT name FROM migrations');
    const appliedMigrations = result.rows.map(row => row.name);
    
    // Run migrations that haven't been applied yet
    let success = true;
    for (const file of files) {
      if (!appliedMigrations.includes(file)) {
        const filePath = path.join(migrationsPath, file);
        const migrationSuccess = await runMigration(filePath);
        
        if (migrationSuccess) {
          // Record successful migration
          await pool.query(
            'INSERT INTO migrations (name) VALUES ($1)',
            [file]
          );
        } else {
          success = false;
          break;
        }
      } else {
        console.log(`Migration ${file} already applied, skipping`);
      }
    }
    
    if (success) {
      console.log('\n✅ All migrations completed successfully');
    } else {
      console.error('\n❌ Some migrations failed');
      process.exit(1);
    }
  } catch (error) {
    console.error('Error running migrations:', error);
    process.exit(1);
  } finally {
    await pool.end();
  }
}

// Run migrations
runMigrations();

File: database/run_migration.py
#!/usr/bin/env python3
"""
Run a specific migration file directly
"""

import os
import sys
import psycopg2
from psycopg2.extras import register_uuid
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Database configuration
DB_CONFIG = {
    'user': os.getenv('DB_USER', 'postgres'),
    'password': os.getenv('DB_PASSWORD', 'vishwak'),
    'host': os.getenv('DB_HOST', 'localhost'),
    'port': os.getenv('DB_PORT', '5432'),
    'database': os.getenv('DB_NAME', 'property_mapping')
}

def run_migration(migration_file):
    """Run a specific migration file"""
    try:
        # Connect to database
        conn = psycopg2.connect(
            user=DB_CONFIG['user'],
            password=DB_CONFIG['password'],
            host=DB_CONFIG['host'],
            port=DB_CONFIG['port'],
            database=DB_CONFIG['database']
        )
        
        # Register UUID type
        register_uuid()
        
        # Create cursor
        cursor = conn.cursor()
        
        print(f"Connected to database: {DB_CONFIG['database']}")
        
        # Read migration file
        with open(migration_file, 'r') as f:
            migration_sql = f.read()
        
        # Execute migration
        cursor.execute(migration_sql)
        
        # Record migration in migrations table
        cursor.execute(
            "INSERT INTO migrations (name) VALUES (%s)",
            (os.path.basename(migration_file),)
        )
        
        # Commit changes
        conn.commit()
        
        print(f"Migration {os.path.basename(migration_file)} applied successfully")
        
        # Close connection
        cursor.close()
        conn.close()
        print("Database connection closed")
        
        return True
    except psycopg2.Error as e:
        print(f"Error running migration: {e}")
        return False

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python run_migration.py <migration_file>")
        sys.exit(1)
    
    migration_file = sys.argv[1]
    
    if not os.path.exists(migration_file):
        print(f"Migration file not found: {migration_file}")
        sys.exit(1)
    
    success = run_migration(migration_file)
    
    if not success:
        sys.exit(1)